### ISSUE TITLE
Contact selection behave weirdly

### ISSUE THREAD

**Issue by devcrafting** [nextcloud/mail#1973]
### Expected behavior
I type an email address, it should take it, but I can't.

### Actual behavior
When I start a new message, I click in "To", I have a "list is empty". I start typing email address, I see a + sign with what I am typing. If I slow down just a bit, it stops keeping new characters (eg I typed "clement@xxx.com", but it keeps only "+ cle" (an invalid address). From this point, I cannot add any address (even with copy/paste)...also the suggestion list always display contacts and an empty entry...

I got several errors in the browser console both in Firefox and Chrome :

```
background sync failed Error: "Request failed with status code 500"
    exports https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:87
    exports https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:94
    onreadystatechange https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:87
mail.js:94:13808
```

and only in Firefox

```
TypeError: "t.customLabel(...) is null"
    optionKeys https://domain.com/custom_apps/mail/js/mail.2.a6edd5281a08a69e52fe.js:470
    optionKeys https://domain.com/custom_apps/mail/js/mail.2.a6edd5281a08a69e52fe.js:470
    get https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:15
    evaluate https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:15
    bn https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:15
    isExistingOption https://domain.com/custom_apps/mail/js/mail.2.a6edd5281a08a69e52fe.js:470
    filteredOptions https://domain.com/custom_apps/mail/js/mail.2.a6edd5281a08a69e52fe.js:470
    get https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:15
    evaluate https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:15
    bn https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:15
    render https://domain.com/custom_apps/mail/js/mail.2.a6edd5281a08a69e52fe.js:470
    _render https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:15
    r https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:15
    get https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:15
    run https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:15
    pn https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:15
    ee https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:15
    Yt https://domain.com/custom_apps/mail/js/mail.js?v=7efbea2b-0:15
mail.js:15:11622
```

and only in Chrome :

```
mail.js?v=7efbea2b-0:15 TypeError: Cannot read property 'toString' of null
    at mail.2.a6edd5281a08a69e52fe.js:470
    at Array.map (<anonymous>)
    at a.optionKeys (mail.2.a6edd5281a08a69e52fe.js:470)
    at hn.get (mail.js?v=7efbea2b-0:15)
    at hn.evaluate (mail.js?v=7efbea2b-0:15)
    at a.optionKeys (mail.js?v=7efbea2b-0:15)
    at a.isExistingOption (mail.2.a6edd5281a08a69e52fe.js:470)
    at a.filteredOptions (mail.2.a6edd5281a08a69e52fe.js:470)
    at hn.get (mail.js?v=7efbea2b-0:15)
    at hn.evaluate (mail.js?v=7efbea2b-0:15)
```

### Mail app

**Mail app version:** : 0.16.0

**Mailserver or service:** : Gmail with 2-factor auth, so using an AppPassword

### Server configuration

**Operating system**: CoreOS + docker-compose from https://github.com/nextcloud/docker/tree/master/.examples/docker-compose/with-nginx-proxy/postgres/fpm

**Web server:** Nginx

**Database:** PostgreSQL

**PHP version:** from nextcloud docker

**Nextcloud Version:** 16.0.4

#### Client configuration
**Browser:** tried with Firefox 68.0.2 & Chrome 76.0.3809.100

**Operating system:** Windows 10

**Comment by elsonjunio** [comment]
I'm having the same problem.

**Mail app**
**Mail app version:** 0.16.0

**Mailserver or service:**  Local Server (Dovecot/Postfix/MS Active Directory)

**Server configuration**
**Operating system:** Debian 9

**Web server:**  Apache/2.4.25 (Debian)

**Database:** MariaDB 10.1.38-MariaDB-0+deb9u1

**PHP version:** PHP 7.2.22-1+0~20190902.26+debian9~1.gbpd64eb7 

**Nextcloud Version:** 16.0.4

**Client configuration**
**Browser:** tried with Firefox 60.0.1 & Chrome 73.0.3683.86 (Versão oficial) 64 bits

**Operating system:** Deepin 15

**Comment by luvis** [comment]
I'm having the same problem, 
Ubuntu 18.04 lts
MariaDB 10
Php 7.2
Mail app and nextcloud latest stable

**Comment by szaimen** [comment]
+1

**Comment by ChristophWurst** [comment]
>  I have to mouse-click the plus-sign.

pro tip: <kbd>Enter</kbd>

**Comment by luvis** [comment]
> > I have to mouse-click the plus-sign.
> 
> pro tip: Enter

Thanks for the tip, however this doesn't work for me. If I type an email address and press enter, nothing happens (most of the time), the adress doesn't get checked. I tried this several times and my success rate, typing the full adress and pressing enter and it actually working, is about 1 out of 4 tries so it's very unreliable.. 
I attached a screenshot of me typing an email adress just to show what it looks like, it only caught the first letter, this is very common btw..

<img width="782" alt="Screenshot 2019-09-16 at 21 39 10" src="https://user-images.githubusercontent.com/619290/64987924-92d5f680-d8ca-11e9-8445-dd4048235b5e.png">

**Comment by ChristophWurst** [comment]
what browser is that?

**Comment by szaimen** [comment]
In my case it's a chromium-based browser.

**Comment by luvis** [comment]
> what browser is that?

Unfortunately I have the same behaviour in Firefox, Chrome, Vivaldi and Safari. It seems though that Safari is just a tad more reliable, however it might be my imagination.. the issue still occurs in all browsers. Firefox is my main browser.

**Comment by luvis** [comment]
@ChristophWurst is this an issue you have been able to replicate or are we just unlucky? Can we do something to fix it? The mail app is in fact un-usable as it is now so I need to either find some other solution or this needs to be fixed. Can I send you some kind of log or info to check? I can let you access my server too if that helps if there's a way to pm on github..

**Comment by AleksovAnry** [comment]
Have the same problem on FF 60.9.0esr Debian 10

**Comment by ChristophWurst** [comment]
Yes, actually I'm indeed unable to reproduce. And we shortly discussed this at our conference but no clue really. It might be the autocomplete callback but not sure.

**Comment by luvis** [comment]
> Yes, actually I'm indeed unable to reproduce. And we shortly discussed this at our conference but no clue really. It might be the autocomplete callback but not sure.

Ok, maybe it only happens with new installs for some reason? I can only speak for myself but all my three installs have the same issue and they are on three different servers with different setups, db and php versions and linux distro. The only thing that's they have in common is that nextcloud and the mail app are freshly installed with the latest version.

**Comment by AleksovAnry** [comment]
Maybe access to my installation helps?

**Comment by ChristophWurst** [comment]
I'd be surprised if it's related to the instance since the code is 100% the same on every installation.

Maybe this is …

* … caused by a specific build of the browser
* … some browser plugin interfering browser events and thus causing delays -> try with a clean profile
* … a slow computer?

But in any case send me the credentials to christoph at nextcloud dot com and I'll try to reproduce.

**Comment by luvis** [comment]
Thanks @ChristophWurst for working on this!

> * … caused by a specific build of the browser
> * … some browser plugin interfering browser events and thus causing delays -> try with a clean profile

I have tried 4 different browsers (Safari, Firefox, Vivaldi, Chrome) all with no plugins or extensions enabled, problem persists.

> * … a slow computer?

Issue is consistent on three different installs, all three are pretty descent. With at least 2.4ghz quad core and 6gb ram. 
My browsing computer is a 2018 macbook pro 15" and I've tried it on a windows pc and an older mac as well.

> 
> But in any case send me the credentials to christoph at nextcloud dot com and I'll try to reproduce.

I sent you the credentials just now, hope you can find something of value.

**Comment by luvis** [comment]
@ChristophWurst 
Just checking in. Where you able to reproduce the issue? I have now made at least 10 fresh installs on three different servers with different Linux Os's as a base and I'm experiencing the exact same behaviour every time (the only similarity on my installs are that I use Virtualmin to set up the database and virtual host, nextcloud is installed by full download and setup through the browser) How is this possible? I'm getting everything else working fine except for the contact selection in the mail app. Very odd behaviour and I'm getting nowhere. Any update would be appreciated. Thanks

@devcrafting just asking, did you get this working somehow or are you still experiencing the issue?

**Comment by ChristophWurst** [comment]
The provided credentials point that a URL that is not a Nextcloud. Can't test right now.

**Comment by luvis** [comment]
> The provided credentials point that a URL that is not a Nextcloud. Can't test right now.

Up and running again, sorry about that, wasn't sure when you'd find the time to check so the server was currently down.

**Comment by ChristophWurst** [comment]
Can't reproduce https://github.com/nextcloud/mail/issues/1973#issuecomment-531927364. I entered that email and it just worked. Didn't stop any any character. Then, after enter,  it got added to the list.

**Comment by luvis** [comment]
> Can't reproduce [#1973 (comment)](https://github.com/nextcloud/mail/issues/1973#issuecomment-531927364). I entered that email and it just worked. Didn't stop any any character. Then, after enter, it got added to the list.

Usually if I type slowly it stops working. let's say I want to write email@email.com and I enter an "email" first and wait two seconds before i keep typing the "@" then the @ symbol won't register even if I was able to type "email" without issue just a second earlier. Same problem when entering more than one email adress. I just tried on the same server and it didn't work.. Please try a few times if you didn't earlier to confirm. If you can get it working perfectly, I'm at a complete loss since I have the same problem from several computers. I even tried a few more computers with other browsers and it really is the same problem I experience..

**Comment by ChristophWurst** [comment]
Tried again. Slower. Same … It's really frustrating. I'm trying my best to reproduce but no luck.

Could you do a screen recording of the browser window? Please open the console and enable XHR logs for the console. Then reproduce. I hope this allows me to understand what's going on here and why I can't trigger it.

**Comment by luvis** [comment]
> Could you do a screen recording of the browser window?

I just sent you showing what it looks like on my end. Hope it gives you something to work with. Thanks Christoph

**Comment by ChristophWurst** [comment]
I can see the problem on the screen recording, but it's impossible to trigger for me. The was also open four literally hours and still everything's smooth.

What OS do you use on the client?

**Comment by luvis** [comment]
> I can see the problem on the screen recording, but it's impossible to trigger for me. The was also open four literally hours and still everything's smooth.
> 
> What OS do you use on the client?

Very strange but thank you for really taking the time. I'm mostly on mac but I've tested on a windows computer as well with the same experience so I really am confused. In any case, I'm going to try a new approach without setting up my server with virtualmin, that's the only thing I can think of that I haven't tried. I'll let you know if it makes any difference..

**Comment by ChristophWurst** [comment]
Don't waste too much time with the server. The front-end always loads the same code into the browser, no matter what server software is used.

**Comment by szaimen** [comment]
@ChristophWurst with which operating System and Browser combination is it working for you? Maybe this is the point?

**Comment by ChristophWurst** [comment]
FF69.0.3 on Arch Linux.

**Comment by szaimen** [comment]
> FF69.0.3 on Arch Linux.

So maybe you could try to reproduce the error on another client with macos or windows?
With the credentials @luvis provided to you?
For me it is not working on windows and with a chromium based browser (btw: I am using nextcloud-snap).

**Comment by robbytobby** [comment]
I did have that exact same issue with a new install of nextcloud 17 and any browser. 
**but**: The issue is gone after I added a first contact to my contacts.

**Comment by luvis** [comment]
> I did have that exact same issue with a new install of nextcloud 17 and any browser.
> **but**: The issue is gone after I added a first contact to my contacts.

Thank you good sir. I tried adding a contact to my contact list and the problem is gone for me as well. I can't believe the fix was so simple =) You just made my day @robbytobby 

@ChristophWurst does this help you in any way to figure out the root of the problem? It now works for me in all my browsers and computers, problem solved.

**Comment by ChristophWurst** [comment]
Okay, so. It only appears if you do not have *any* contacts?

**Comment by luvis** [comment]
> Okay, so. It only appears if you do not have _any_ contacts?

Rather, the issue occurred before adding a first contact. I tried deleting the contact again to see if the problem would return, but it did not.

**Comment by jancborchardt** [comment]
cc'ing @skjnldsv for Contacts then and @rullzer generally. As on a new install people generally have no contacts, this case should be caught?

**Comment by ChristophWurst** [comment]
On https://help.nextcloud.com/t/mail-app-sending-mail-hangs-in-chromium/66006/14 I read this can be narrowed down to be a problem with Chromium but not Chrome.

**Comment by benkees** [comment]
I can confirm now, that the buggy behavior of the auto completion doesn't occur with NC 18 with Mail 1.1.2 with the same browser and os (client site).

**Comment by ahstro** [comment]
We might have a fix for this in #2703. If anyone who has experienced it knows how to test code, that might confirm the hunch 😇

**Comment by ChristophWurst** [comment]
Re-reading the original report, I'm convinced #2703 will fix the bug. Thanks @ahstro!
